_posts/2024/2024-02-06-about-caching-in-eo.md

```diff
@@ -0,0 +1,283 @@
+Due to this issue, the program behaved incorrectly, because saving the goal's result to the cache is not instantaneous.
+After conducting an in-depth analysis of the project's incorrect operation,
+several disadvantages of the previous EO caching mechanism were brought to light:
+* Incorrect search conditions for data in the cache.
+* The verification method requires reading the file content, which results in inefficiencies.
+* The presence of multiple caching mechanisms creates challenges in identifying and rectifying caching errors.
+* Employing multiple caching mechanisms for similar entities is a suboptimal practice, 
+leading to redundancy and complicating the caching infrastructure.
+
+
+In tackling caching challenges within EO, we conducted a thorough evaluation of current caching systems.
+Most existing caching systems are not suitable for the EO project.
+However, one candidate emerged as a potential solution for caching EO compilation stages: the Takari Incremental API.
+The Takari Incremental API exhibits key similarities with the EO caching system,
+notably in its utilization of inputs and outputs directories, absence of a hash for data storage and retrieval,
+and compatibility with Maven goals.
+However, it diverges from the EO caching approach in one significant aspect – the absence of a distinct cache directory.
+
+We can try to use this API or implement our own caching approach, correcting the disadvantages found.
+The envisioned approach involves the creation of a singular class responsible
+for storing and retrieving data from the cache.
+The logic for checking the relevance of cached data is presented below:
+1) We create EO program, named "example".
+   Intermediate files during compilation of this program will have the same name, but not the format
+   (e.g. `example.eo`, `example.xml`).
+2) When the EO compiler compiles this program task, it saves intermediate files of compilation steps into cache.
+   Each compilation step has own caching directory.
+3) When the EO compiler starts a project build again, it will check if there is a file, named "example",
+   in the cache of each step. If such a file exists,
+   then it is enough to check that the last modification time of this file at the current step
+   is later than at the previous step. If this condition is true,
+   then the finished file can be retrieved from the cache.
+   Below is a diagram illustrating the EO compilation steps, which have caching directory for EO version `0.34.0`:
+
+<p align="center">
+  <img src="/images/SavingInCacheEO.svg">
+</p>
+
+4) If the EO program file [Picture 5](/images/RewritingInCacheEO1.svg) 
+   or an intermediate file [Picture 6](/images/RewritingInCacheEO2.svg) have changed,
+   then the previously cached files becomes invalid.
+   In this case, the compilation step performs an action again and rewrites outputs.
+
+<p align="center">
+  <img src="/images/RewritingInCacheEO1.svg">
+</p>
+
+
+<p align="center">
```

Comment:
@Yanich96, I agree that we need to remove the redundancy in the code. For example, we should combine the `Footprint` and `Optimization` methods and fix the hash comparison mechanism. As for checking the "previous" step, I completely disagree. First of all, some steps might be skipped, and more importantly, doing this significantly increases [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming)) between the phases, which is a significant architectural flaw.

Reply:
@Yanich96 [This](https://github.com/objectionary/news.eolang.org/pull/58/files#r1602623769) comment is still relevant.